packages/engine/lib/error/src/macros.rs

```diff
@@ -1,62 +1,99 @@
 #[doc(hidden)]
 pub mod __private {
-    #![allow(clippy::unused_self)]
-    //! Uses [autoref-based stable specialization](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md).
-    // TODO: Expand documentation when string literals are forbidden as this will shrink the
-    //   implementation by a fair bit.
+    //! Implementation detail for macros.
+    //!
+    //! ⚠️ **Functionality in this module is considered unstable and is subject to change at any time
+    //! without a major version bump!** ⚠️
+    mod specialization {
+        #![allow(clippy::unused_self)]
+        //! [Autoref-Based Stable Specialization](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md)
+        //! for macros.
+        //!
+        //! This is a stable implementation for specialization (only possible within macros, as
+        //! there is no trait bound for this things).
+        //!
+        //! The different tags [`ReportTag`], [`ErrorTag`], and [`ContextTag`] have a blanket
+        //! implementation returning a concrete type. This type is then used to create a [`Report`].
+        //!
+        //! [`ErrorTag`] is implemented for `T: `[`Error`]s, [`ContextTag`] is implemented for `&T:
+        //! `[`Context`]s. As [`ContextTag`] is implemented references, [`ErrorTag`] has a higher
+        //! precedence and will be picked up first. So calling `my_error.kind()` will always
+        //! [`ErrorReporter`]. This can then be used to create a [`Report`] by calling
```

Comment:
Or are you describing the return type, so 'will always return'?
```suggestion
        //! will always return [`ErrorReporter`]. This can then be used to create a [`Report`] by calling
```

Reply:
The latter 😆